src/python/dxpy/bindings/dxfile_functions.py

```diff
@@ -466,3 +466,101 @@ def upload_string(to_upload, media_type=None, keep_open=False, wait_on_close=Fal
+        project_folders = list_subfolders.folders_cache[project]
+    else:
+        project_folders = dxpy.get_handler(project).describe(input_params={'folders': True})['folders']
+    # TODO: support shell-style path globbing (i.e. /a*/c matches /ab/c but not /a/b/c)
+    # return pathmatch.filter(project_folders, os.path.join(path, '*'))
+    if recurse:
+        return (f for f in project_folders if f.startswith(path))
+    else:
+        return (f for f in project_folders if f.startswith(path) and '/' not in f[len(path)+1:])
+
+def download_folder(project, destdir, folder="/", overwrite=False, chunksize=dxfile.DEFAULT_BUFFER_SIZE,
+        usecache=True, **kwargs):
+    '''
+    :param project: Project ID to use as context for this download.
+    :type project: string
+    :param destdir: Local destination location
+    :type destdir: string
+    :param folder: Path to the remote folder to download
+    :type folder: string
+    :param overwrite: Overwrite existing files
+    :type overwrite: boolean
+    :param usecache: Use folders listing cache
+    :type usecache: boolean
+
+    Downloads a remote *folder* of the *project* and saves it to the *destdir* local location.
+
+    Example::
+
+        download_folder("project-xxxx", "/home/jsmith/input", folder="/input")
+
+    '''
+
+    def ensure_local_dir(d):
+        if not os.path.isdir(d):
+            if os.path.exists(d):
+                raise DXFileError("Destination location '{}' already exists and is not a directory".format(destdir))
+            logger.debug("Creating destination directory: '%s'", d)
+            os.makedirs(d)
+
+    def compose_local_dir(remote_subfolder):
+        return os.path.normpath(os.path.join(destdir, remote_subfolder[1:] if folder == "/" else remote_subfolder[len(folder) + 1:]))
+
+    folder = os.path.normpath(folder).strip()
+    if folder == "":
+        raise DXFileError("Invalid remote folder name: '{}'".format(folder))
+    destdir = os.path.normpath(destdir).strip()
+    if destdir == "":
+        raise DXFileError("Invalid destination directory name: '{}'".format(destdir))
+
+    remote_folders = list_subfolders(project, folder, usecache=usecache, recurse=True)
+
+    # Creating target directory tree
+    remote_subfolders = [f for f in remote_folders if f.startswith(folder)]
+    if len(remote_subfolders) <= 0:
+        raise DXFileError("Remote folder '{}' not found".format(folder))
+    remote_subfolders.sort()
+    for remote_subfolder in remote_subfolders:
+        ensure_local_dir(compose_local_dir(remote_subfolder))
+
+    # Downloading files
+    for remote_file in dxpy.search.find_data_objects(classname='file', state='closed', project=project, folder=folder,
+            recurse=True, describe=True):
+        local_filename = os.path.join(compose_local_dir(remote_file['describe']['folder']), remote_file['describe']['name'])
```

Comment:
Again, I wonder if something special needs to be done to transform the unity path `folder` into a (possibly-Windows) local filename... if you aren't able to test this in action, please leave a comment saying that the behavior of this code on Windows needs to be looked at some more.